projects/plugins/crm/includes/ZeroBSCRM.DAL3.Obj.Invoices.php

```diff
@@ -2753,52 +2753,57 @@ public function getOutstandingBalance($invoiceID=-1){
-                // this one'll be a rolling sum
-                $transactionsTotalValue = 0.0;
 
-                // cycle through trans + calc existing balance
-                if (isset($invoice['transactions']) && is_array($invoice['transactions'])){
+						// this one'll be a rolling sum
+						$transactions_total_value = 0.0;
 
-                    // got trans
-                    foreach ($invoice['transactions'] as $transaction){
+						// cycle through trans + calc existing balance
+						if ( isset( $invoice['transactions'] ) && is_array( $invoice['transactions'] ) ) {
 
-                        // should we also check for status=completed/succeeded? (leaving for now, will let check all):
+							// got trans
+							foreach ( $invoice['transactions'] as $transaction ) {
 
-                        // get amount
-                        $transactionAmount = 0.0; if (isset($transaction['total'])) $transactionAmount = (float)$transaction['total'];
+								// should we also check for status=completed/succeeded? (leaving for now, will let check all):
 
-                        if ($transactionAmount > 0){
+								// get amount
+								$transaction_amount = 0.0;
 
-                            switch ($transaction['type']){
-
-                                case __('Sale','zero-bs-crm'):
-
-                                    // these count as debits against invoice.
-                                    $transactionsTotalValue -= $transactionAmount;
-
-                                    break;
-
-                                case __('Refund','zero-bs-crm'):
-                                case __('Credit Note','zero-bs-crm'):
+								if ( isset( $transaction['total'] ) ) {
+									$transaction_amount = (float) $transaction['total'];
+								}
 
-                                    // these count as credits against invoice.
-                                    $transactionsTotalValue += $transactionAmount;
+								switch ( $transaction['type'] ) {
 
-                                    break;
+									case __( 'Sale', 'zero-bs-crm' ):
+										// these count as debits against invoice.
+										$transactions_total_value -= $transaction_amount;
 
+										break;
 
+									case __( 'Refund', 'zero-bs-crm' ):
+									case __( 'Credit Note', 'zero-bs-crm' ):
+										// these count as credits against invoice, but should be marked as negative (thus being deducted).
+										$transactions_total_value += $transaction_amount;
+										// These count as credits against invoice, and should be added.
+										// If inputted as negative, then add that amount.
+										if ( str_starts_with( $transaction_amount, '-' ) ) {
+											$transactions_total_value += $transaction_amount;
+										} else {
+											// If inputted as positive, then subtract that amount.
+											$transactions_total_value -= $transaction_amount;
+										}
```

Comment:
I don't understand this part at all...

if the invoice amount is 100 and the transaction is a Refund or a Credit Note with the amount of -30:

- `$invoiceTotalValue = 100.00`
- `$transaction_total_value = 0.0`
- `$transaction_amount = -30` initial amount
- `$transaction_total_value += $transaction_amount` => -30
- `$transaction_amount` is negative so `$transaction_total_value += $transaction_amount` => `$transaction_total_value = -60`
- Value to return: `$invoiceTotalValue + $transaction_total_value` = 40

If the transaction is positive, the `$transaction_total_value` will be `0` so the result to return is `100`.

Reply:
Oh good catch. That extra line - `$transactions_total_value += $transaction_amount;` - was left in from the original code and should have been removed. I've taken it out now: 9267a249f0587ab46097ffbb7af52576d284c7ff

For me part of the challenge with that section of code is that I'm not sure how to actually test it so that's part of why I didn't catch that - but as the functionality here should match elsewhere it should be matching any changes elsewhere.